### PR TITLE
[fix]: [SRM-13994]: and SRM-13795 reduce the retry counts for deploym…

### PR DESCRIPTION
**Pull request by AbhijithHarness**
…ent data collection tasks and handle timeouts

### Old behaviour:
  1. The data collection execution used to have a timeout of 3 minutes, and if this is breached, the entire PT used to fail with exception.
  2. The status and exception was not propagated to DCT.

### New Behaviour
1. PT doesnt fail, if DC is timed out.
2. The Failed status and exception was not propagated to DCT.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)

### CODE REVIEW

**Review by dhruv-harness — COMMENTED** [review]
Can we add some tests for the new behaviour?

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165#-:-NOpDQld6Pxd25MdLBNg:b-jbgwtl)* status: 0 of 1 approvals obtained. Waiting on team(s): del-team. 

___
*[`260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java` line 131 at r1](https://reviewable.io/reviews/harness/harness-core/44165#-NOpD6p5DWfk0sCHuJwA:-NOpD6p5DWfk0sCHuJwB:b-i9nc3y) ([raw file](https://github.com/harness/harness-core/blob/ca401ab59b42eb9a223a5e2a4c3fc71d5bbe554e/260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java#L131)):*
> ```Java
>                                            -> run(taskParams, dataCollectionInfo.getConnectorConfigDTO(), dct),
>                                  parallelExecutor)
>                              .orTimeout(3, TimeUnit.MINUTES)
> ```

nit: We can define a constant for this timeout for better readability. Also do we want to keep it longer than 1 min?

File: 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java [hunk 1]
```diff
@@ -121,12 +121,26 @@ public PerpetualTaskResponse runOnce(
...
         } else {
           log.info("Next tasks to process: {}", dataCollectionTasks);
-          List<Callable<Void>> callables = new ArrayList<>();
-          dataCollectionTasks.forEach(dataCollectionTaskDTO -> callables.add(() -> {
-            run(taskParams, dataCollectionInfo.getConnectorConfigDTO(), dataCollectionTaskDTO);
-            return null;
-          }));
-          cvngParallelExecutor.executeParallel(callables);
+          List<CompletableFuture> completableFutures =
+              dataCollectionTasks.stream()
```
Comment (n00bitax) [inline 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java:125]: nit : is it null safe?

File: 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java [hunk 2]
```diff
@@ -121,12 +121,26 @@ public PerpetualTaskResponse runOnce(
...
+                             .orTimeout(3, TimeUnit.MINUTES)
+                             .exceptionally(exception -> {
+                               updateStatusWithException(taskParams, dct, exception);
+                               return null;
+                             }))
+                  .collect(Collectors.toList());
+          // execute and wait for all CFs to complete parallely
+          CompletableFuture.allOf(completableFutures.toArray(n -> new CompletableFuture[n]))
+              .exceptionally(ex -> {
+                log.error("Exception in parrallel execution:", ex);
```
Comment (n00bitax) [inline 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java:140]: nit : small typo parallel, also we can name the variable something other than n.

**Review by n00bitax — APPROVED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: Waiting on team(s): del-team.

**Review by AbhijithHarness — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165#-:-NPQffs9D37MAWV6GS6d:bb7f2mf)* status: 2 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained. Waiting on team(s): del-team. 

___
*[`260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java` line 125 at r1](https://reviewable.io/reviews/harness/harness-core/44165#-NOo6bfT6H9qdvqWOk-8-r1-125:-NPQehGNBYY7QxJ6wNSp:b-xdfunm) ([raw file](https://github.com/harness/harness-core/blob/ca401ab59b42eb9a223a5e2a4c3fc71d5bbe554e/260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java#L125)):*
<details><summary><i>Previously, n00bitax (Ansuman Satapathy) wrote…</i></summary><blockquote>

nit : is it null safe?
</blockquote></details>

Null check is present above
___
*[`260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java` line 131 at r1](https://reviewable.io/reviews/harness/harness-core/44165#-NOpD6p5DWfk0sCHuJwA:-NPQeofy7OKb2cKbQftc:b-71m25q) ([raw file](https://github.com/harness/harness-core/blob/ca401ab59b42eb9a223a5e2a4c3fc71d5bbe554e/260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java#L131)):*
<details><summary><i>Previously, dhruv-harness (Dhruv) wrote…</i></summary><blockquote>

nit: We can define a constant for this timeout for better readability. Also do we want to keep it longer than 1 min?
</blockquote></details>

Moved it out for test case
___
*[`260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java` line 140 at r1](https://reviewable.io/reviews/harness/harness-core/44165#-NOo6bfT6H9qdvqWOk-8-r1-140:-NPQf58l8IGwkI_ZyogM:b-7vw3yj) ([raw file](https://github.com/harness/harness-core/blob/ca401ab59b42eb9a223a5e2a4c3fc71d5bbe554e/260-delegate/src/main/java/io/harness/perpetualtask/datacollection/DataCollectionPerpetualTaskExecutor.java#L140)):*
<details><summary><i>Previously, n00bitax (Ansuman Satapathy) wrote…</i></summary><blockquote>

nit : small typo parallel, also we can name the variable something other than n.
</blockquote></details>

it is just length, n feels fine for that.,

**Review by deepak-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Review by dhruv-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Review by dhruv-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Review by AbhijithHarness — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: Waiting on team(s): del-team.

**Review by n00bitax — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: Waiting on team(s): del-team.

**Comment by bharatgoelharness** [comment]
Please add a detail description and test cases which you have tested in PR env. cc @raghuAtWings

**Review by jennyjames21 — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44165)* status: Waiting on team(s): del-team.

**Pull request merged**